File: android/src/main/java/com/mqtt/utils/MqttEventEmitter.kt
package com.mqtt.utils

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.mqtt.models.events.MqttEvent
import com.mqtt.models.events.MqttEventParam
import org.eclipse.paho.client.mqttv3.MqttException

class MqttEventEmitter(private val reactContext: ReactContext, private val clientRef: String) {
  /**
   * forwards a raised exception downstream to the RN event emitter.
   * @param e The exception to forward.
   */
  fun forwardException(e: Throwable) {
    val params = Arguments.createMap()
    if (e is MqttException) {
      params.putInt(MqttEventParam.ERR_CODE.name, e.reasonCode)

      if (e.reasonCode == 0 && e.cause != null) return forwardException(e.cause!!)
    }
    params.putString(MqttEventParam.ERR_MESSAGE.name, e.message)
    params.putString(MqttEventParam.STACKTRACE.name, e.stackTrace.joinToString("\n\t") {
      "${it.fileName} - ${it.className}.${it.methodName}:${it.lineNumber}"
    })
    sendEvent(MqttEvent.EXCEPTION, params)
  }

  /**
   * sends an event to the RN event emitter.
   * @param event The event to report.
   * @param params The parameters to send with the event.
   */
  fun sendEvent(event: MqttEvent, params: WritableMap = Arguments.createMap()) {
    if (reactContext.hasActiveReactInstance()) {
      params.putString(MqttEventParam.CLIENT_REF.name, clientRef)
      reactContext
        .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
        .emit(event.name, params)
    }
  }
}

File: ios/MqttEventEmitter.swift
import Foundation

class MqttEventEmitter {
  public static var shard: RCTEventEmitter!
  private let clientRef: String

  init(_ eventEmitter: RCTEventEmitter, _ clientRef: String) {
    self.clientRef = clientRef
    MqttEventEmitter.shard = eventEmitter
  }

  func forwardException(e: Error) {
    let params: [String: Any] = [
      MqttEventParam.ERR_CODE.rawValue: 0,
      MqttEventParam.ERR_MESSAGE.rawValue: e.localizedDescription,
      MqttEventParam.STACKTRACE.rawValue: ""
    ]
    if Mqtt.hasListener {
      MqttEventEmitter.shard.sendEvent(withName: MqttEvent.EXCEPTION.rawValue, body: params)
    }
  }

  func sendEvent(event: MqttEvent, params: NSMutableDictionary = [:]) {
    if Mqtt.hasListener {
      params.setValue(self.clientRef, forKey: MqttEventParam.CLIENT_REF.rawValue)
      MqttEventEmitter.shard.sendEvent(withName: event.rawValue, body: params)
    }
  }
}
